speculation-rules.bs

```diff
@@ -317,47 +318,51 @@ add the following step
+  To <dfn>parse speculation rules</dfn> given a [=string=] |input|, [=document=] |document|, and a [=URL=] |baseURL|, perform the following steps. They return a [=speculation rule set=] or null.
 
   1. Let |parsed| be the result of [=parsing a JSON string to an Infra value=] given |input|.
   1. If |parsed| is not a [=map=], then return null.
   1. Let |result| be an empty [=speculation rule set=].
   1. If |parsed|["`prefetch`"] [=map/exists=] and is a [=list=], then [=list/for each=] |prefetchRule| of |parsed|["`prefetch`"]:
     1. If |prefetchRule| is not a [=map=], then [=iteration/continue=].
-    1. Let |rule| be the result of [=parsing a speculation rule=] given |prefetchRule| and |baseURL|.
+    1. Let |rule| be the result of [=parsing a speculation rule=] given |prefetchRule|, |document| and |baseURL|.
     1. If |rule| is null, then [=iteration/continue=].
     1. If |rule|'s [=speculation rule/target navigable name hint=] is not null, then [=iteration/continue=].
     1. [=list/Append=] |rule| to |result|'s [=speculation rule set/prefetch rules=].
   1. If |parsed|["`prerender`"] [=map/exists=] and is a [=list=], then [=list/for each=] |prerenderRule| of |parsed|["`prerender`"]:
     1. If |prerenderRule| is not a [=map=], then [=iteration/continue=].
-    1. Let |rule| be the result of [=parsing a speculation rule=] given |prerenderRule| and |baseURL|.
+    1. Let |rule| be the result of [=parsing a speculation rule=] given |prerenderRule|, |document|, and |baseURL|.
     1. If |rule| is null, then [=iteration/continue=].
     1. [=list/Append=] |rule| to |result|'s [=speculation rule set/prerender rules=].
   1. Return |result|.
 </div>
 
 <div algorithm="parse a speculation rule">
-  To <dfn>parse a speculation rule</dfn> given a [=map=] |input| and a [=URL=] |baseURL|, perform the following steps. They return a [=speculation rule=] or null.
+  To <dfn>parse a speculation rule</dfn> given a [=map=] |input|, a [=document=] |document|, and a [=URL=] |baseURL|, perform the following steps. They return a [=speculation rule=] or null.
 
-  1. If |input| has any [=map/key=] other than "`source`", "`urls`", "`where`", "`requires`", and `"target_hint"`, then return null.
+  1. If |input| has any [=map/key=] other than "`source`", "`urls`", "`where`", "`requires`", `"target_hint"`, and "`relative_to`", then return null.
   1. If |input|["`source`"] does not [=map/exist=] or is neither the [=string=] "`list`" nor the [=string=] "`document`", then return null.
   1. Let |source| by |input|["`source`"].
   1. Let |urls| be an empty [=list=].
   1. Let |predicate| be null.
   1. If |source| is "`list`":
     1. If |input|["`where`"] [=map/exists=], then return null.
     1. If |input|["`urls`"] does not [=map/exist=], is not a [=list=], or has any element which is not a [=string=], then return null.
+    1. If |input|["`relative_to`"] [=map/exists=]:
+      1. If |input|["`relative_to`"] is not "`document`", then return null.
```

Comment:
It might be nice to have a value representing the default, e.g. `"relative_to": "speculation-rules"`. It wouldn't really be useful for anything besides documentation and teaching purposes.

Reply:
I added in a default value; I called it "ruleset" instead as we use the ruleset source's base URL (but I don't feel super strongly about this an can use "speculation-rules" if you think that is better).

Reply:
LGTM.

I am unsure how I feel about "ruleset" since that term isn't exposed to web developers in any other place; I thought "speculation-rules" (or "speculationrules"??) might work because that's probably the most familiar term for "what goes inside a `<script type="speculationrules">`" and "what's inside the document at the other end of `Speculation-Rules:`".

But, I am happy with either, and this would be easy to change later (probably nobody will depend on this in any significant way). I'll let you and @jeremyroman make the call in your timezone. Please feel free to merge without further review from me :)

Reply:
Heh, the reason I didn't include this in my original proposal was exactly because I wasn't sure what to call it.

I think "ruleset" would be my mild preference (because it's relative to this particular ruleset, not all speculation rules globally) but this is only mild; "speculation-rules" totally works too (and @domenic's point is valid).

Reply:
Jeremy's rationale is why I didn't like "speculation-rules" as well. We do have the notion of a rule set in the specification, and it's also mention a few times in the explainer so it's not a complete unknown IMO.